flang/lib/Lower/IO.cpp

```diff
@@ -977,9 +977,9 @@ static void genIoLoop(Fortran::lower::AbstractConverter &converter,
     fir::StoreOp::create(builder, loc, lcv, loopVar);
     genItemList(ioImpliedDo);
     builder.setInsertionPointToEnd(doLoopOp.getBody());
-    mlir::Value result = mlir::arith::AddIOp::create(
-        builder, loc, doLoopOp.getInductionVar(), doLoopOp.getStep(), iofAttr);
-    fir::ResultOp::create(builder, loc, result);
+    // fir.do_loop's induction variable's increment is implied,
+    // so we do not need to increment it explicitly.
+    fir::ResultOp::create(builder, loc, doLoopOp.getInductionVar());
```

Comment:
Do you think it would be possible to align the IO case with the normal do Loop case and have a loop argument/result with the user type + explicit increment and `finalCountValue=false`?

It seems to be the only remaining case where `finalCountValue=true`, and I wonder if it would not be best to just remove this attribute if it is not really needed to have a more uniform and simpler IR (no need to do the removal in this patch that is already touching many files).

Reply:
I thought about that, and was not sure if limiting `fir.do_loop` that way is reasonable. I suppose it should be okay to use the final value of the counter, in general. For example, [HL]FIR optimization passes may want to generate such a `fir.do_loop` without introducing another loop-carried value. Of course, the final value may be computed as a function of the lower and upper bound of the loop, so we can indeed limit the operation to never return the final counter.

I have no strong opinion about this.

Reply:
No strong opinion easier.